spec/function.dd

```diff
@@ -173,124 +173,246 @@ int foo();
-        )
+$(P A pure function can throw exceptions (unless it has the `nothrow` attribute).)
 
-        $(P A pure function can throw exceptions.)
+$(P Note: The rules above do not preclude modification of data obtained through parameters. Unlike
+traditional functional programming languages, D's pure functions allow modification of the caller's
+state through their mutable parameters. This definition of mutable functions is more relaxed than
+the one traditionally employed by pure functional languages because it allows a pure function to
+use state mutation, as long as all state is created internally or reachable through its arguments.
+In particular, a pure function may allocate memory by means of e.g. `new` or `malloc` without these
+being special cases. Example:)
 
 ---
-import std.stdio;
-int x;
-immutable int y;
-const int* pz;
-
-pure int foo(int i,
-             char* p,
-             const char* q,
-             immutable int* s)
-{
-    debug writeln("in foo()"); // ok, impure code allowed in debug statement
-    x = i;   // error, modifying global state
-    i = x;   // error, reading mutable global state
-    i = y;   // ok, reading immutable global state
-    i = *pz; // error, reading const global state
-    return i;
+pure size_t foo(int[] arr) { arr[] += 1; return arr.length; }
+int[] a = [1, 2, 3];
+foo(a);
+assert(a == [2, 3, 4]);
+---
+
+$(P Pure functions fall in exactly one of two categories: $(I weakly pure) and $(I strongly pure).
+There are no dedicated annotations for these; instead, a set of rules applied to the function
+declaration determines its category.)
+
+$(H3 Weak Purity)
+
+$(P A function annotated with `pure` is weakly pure if and only if at least one of the
+following conditions are met:)
+
+$(OL
+$(LI One or more parameter types are incompletely defined. Example:)
+
+---
+struct S; // defined elsewhere
+// These functions are weakly pure:
+pure int fun(S object);
+pure int fun(immutable S object);
+pure int fun(immutable S* pointer);
+---
+
+$(LI One or more parameters allow indirect access$(MDASH)through `ref`, class reference, dynamic
+array, or pointer$(MDASH)to unqualified data or data with qualifiers other than `immutable`.
+Example:)
+
+---
+// These functions are weakly pure:
+pure int fun(Object obj);
+pure int fun(int* x);
+pure int fun(const int[] x);
+struct S
+{
+    const int x;
+}
+pure int fun(S* object);
+---
+
+$(LI The function returns `void`. Example:)
+
+---
+// These functions are weakly pure:
+pure void fun(int);
+pure void fun(immutable int* x);
+---
+
+$(LI The returned type allows indirect access$(MDASH)through `ref`, class reference, dynamic
+array, or pointer$(MDASH)to unqualified data or data with qualifiers other than `immutable`.
+Example:)
+
+---
+// These functions are weakly pure:
+pure Object fun(int);
+pure int* fun(int);
+pure const(int[]) fun(int);
+struct S
+{
+    const int x;
+}
+pure S* fun(int);
+---
+
+$(LI The function takes zero arguments. Example:)
```

Comment:
Why?

Reply:
Such functions are already accepted and considering them strongly pure would be useful - they'd be constants. So I put them in the weakly pure category.

Weakly pure means "somewhat restricted, can be called by any pure function, but not reordered at will".

Reply:
Constants or factory functions.

Reply:
@tgehr cool! So those should be weakly pure, correct?

Reply:
No, they should be strongly pure. Note that your pull request says they can't be factory functions.
(Here, I'm ignoring issues like exceptions and termination etc.)